scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/descriptor/JobDescriptorImpl.java

```diff
@@ -132,21 +150,54 @@ public JobDescriptorImpl(InternalJob job) {
     private void makeTree(InternalJob job) {
 
         //create task descriptor list
+
+        Set<String> taskNames = new HashSet<>();
+
         for (InternalTask td : job.getITasks()) {
             //if this task is a first task, put it in eligible tasks list
             EligibleTaskDescriptor lt = new EligibleTaskDescriptorImpl(td);
 
-            if (isEntryPoint(td, job.getITasks())) {
+
+            List<TaskState> deps = td.getDependences();
+
+            if (deps == null || deps.size() == 0) {
                 eligibleTasks.put(td.getId(), lt);
             }
 
+            FlowScript sc = td.getFlowScript();
+            if (sc != null) {
+                String actionType = sc.getActionType();
+                if (FlowActionType.parse(actionType).equals(FlowActionType.IF)) {
+                    String tIf = sc.getActionTarget();
+                    String tElse = sc.getActionTargetElse();
+                    String tJoin = sc.getActionContinuation();
+                    if(tIf != null && !tIf.equals(td.getName())){
+                        taskNames.add(tIf);
+                    }
+                    if(tElse != null && !tElse.equals(td.getName())){
+                        taskNames.add(tElse);
+                    }
+                    if(tJoin != null && !tJoin.equals(td.getName())){
+                        taskNames.add(tJoin);
+                    }
+                }
+            }
+
             if (td.getJoinedBranches() != null || td.getIfBranch() != null) {
                 branchTasks.put(td.getId(), lt);
             }
 
             allTasksWithTheirChildren.put(td, lt);
         }
 
+        final Iterator<Entry<TaskId, EligibleTaskDescriptor>> iterator = eligibleTasks.entrySet().iterator();
+        while(iterator.hasNext()){
+            final Entry<TaskId, EligibleTaskDescriptor> entry = iterator.next();
+            if(taskNames.contains(((EligibleTaskDescriptorImpl)entry.getValue()).getInternal().getName()){
+                iterator.remove();
```

Comment:
can you not simply iterate on the not eleignible task names and use map.remove()